### DISCUSSION TITLE
Flask SocketIO and future retirement of eventlet

### DISCUSSION THREAD

**Discussion by spmsh** [miguelgrinberg/Flask-SocketIO#2037] [General]
Hi @miguelgrinberg, have you seen or what do you think about the warning announcement on the home page of the eventlet project saying that 

> new usages of eventlet are now heavily discouraged

https://eventlet.readthedocs.io/en/latest/
https://eventlet.readthedocs.io/en/latest/migration.html#migration-guide

Since eventlet is the first choice when using Flask SocketIO, what would be the recommendation or change to be done? Any planned update of your Flask SocketIO lib to use asyncio instead of eventlet? What would be your recommendation/suggestion on this?

Thanks.

**Comment by miguelgrinberg** [comment]
Yes, I have seen it. The project is slowly winding down, so at this time I would not recommend its use on new projects.

**Reply by sylos** [reply]
Are you saying flask socketIO is winding down?

**Reply by miguelgrinberg** [reply]
Eventlet is winding down, so you should not use. Flask-SocketIO is fine, use it all you want.

**Comment by spmsh** [comment]
what would you recommend for existing project using flask_socketio ? move to asyncio (although I wouldn't know how), use gevent, or stay with current eventlet version (potentially forever). Thanks again.

**Reply by miguelgrinberg** [reply]
If eventlet works you can continue using it. Else, using regular threading is the safest. Gevent might also be okay. Asyncio is not an option as it isn't compatible with flask.

**Reply by spmsh** [reply]
flask 3.0 still doesn't support asyncio? I've found this in their document: https://flask.palletsprojects.com/en/3.0.x/async-await
thanks,

**Reply by miguelgrinberg** [reply]
Async support in Flask is incomplete at this time. The limitations are well explained in the page that you linked.

**Reply by ssrihari4102** [reply]
> using regular threading is the safest

I'm a bit of a Python noob and trying to get [an old project](https://github.com/rimazmk/hangman-be) working by updating some dependencies. I was wondering what you meant by this and how I should adjust my app from [the existing deployment directions](https://flask-socketio.readthedocs.io/en/latest/deployment.html#gunicorn-web-server) in response?

**Reply by miguelgrinberg** [reply]
This is the third option mentioned in the documentation section that you linked. I need to update that section, the third option with threads is now preferred, and if that does not work then the alternative is to use gevent.

Basically you need to uninstall eventlet and remove any references to it in your application. If you use `async_mode='eventlet'` in your code, remove that as well, and let Flask-SocketIO use the default async mode. Uninstalling eventlet is important, because the default async mode is chosen from what packages are installed. You do not want to have gevent nor eventlet in your virtualenv, so that the threading mode is chosen.
